File: backend/src/main/java/org/example/echo/service/TagService.java
package org.example.echo.service;

import lombok.RequiredArgsConstructor;
import org.example.echo.dto.source.SourceResponse;
import org.example.echo.dto.tag.TagCreateRequest;
import org.example.echo.dto.tag.TagResponse;
import org.example.echo.dto.tag.TagUpdateRequest;
import org.example.echo.entity.Tag;
import org.example.echo.exception.custom.EntityAlreadyExistsException;
import org.example.echo.exception.custom.EntityNotFoundException;
import org.example.echo.mapper.TagMapper;
import org.example.echo.repository.TagRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class TagService {
    private final TagRepository tagRepository;

    public TagResponse createTag(TagCreateRequest request) {
        if (tagRepository.existsByName(request.name()))
            throw new EntityAlreadyExistsException("Tag", request.name());

        Tag tag = Tag.builder()
                .name(request.name())
                .build();

        return TagMapper.toResponse(tagRepository.save(tag));
    }

    public List<TagResponse> getAllTagsList() {
        return tagRepository.findAll().stream()
                .map(TagMapper::toResponse).toList();
    }

    public TagResponse getTagById(Long id) {
        Tag tag = tagRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Tag", id));

        return TagMapper.toResponse(tag);
    }

    @Transactional
    public TagResponse updateTag(Long id, TagUpdateRequest request) {
        Tag tag = tagRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Tag", id));

        tag.setName(request.name());

        return TagMapper.toResponse(tagRepository.save(tag));
    }

    public void deleteTag(Long id) {
        Tag tag = tagRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Tag", id));

        tagRepository.delete(tag);
    }

    public List<TagResponse> getRandomTags(int size) {
        return tagRepository.getRandomTags(size).stream().map(TagMapper::toResponse).toList();
    }

    public List<TagResponse> getFirstTags(int size) {
        return tagRepository.getFirstTags(size).stream().map(TagMapper::toResponse).toList();
    }

    public List<TagResponse> searchTagByName(String q) {
        return tagRepository.searchByName(q).stream().map(TagMapper::toResponse).toList();
    }
}

File: frontend/src/api/tagService.ts
import apiClient from './apiClient';
import { TagResponse, TagCreateRequest, TagUpdateRequest } from '../types';

export const tagService = {
  async getTagById(id: number): Promise<TagResponse> {
    const response = await apiClient.get<TagResponse>(`/tags/${id}`);
    return response.data;
  },

  async searchTags(query: string): Promise<TagResponse[]> {
    const response = await apiClient.get<TagResponse[]>('/tags/search', {
      params: { q: query },
    });
    return response.data;
  },

  async getRandomTags(size: number = 10): Promise<TagResponse[]> {
    const response = await apiClient.get<TagResponse[]>('/tags/random', {
      params: { size },
    });
    return response.data;
  },

  async getFirstTags(size: number = 10): Promise<TagResponse[]> {
    const response = await apiClient.get<TagResponse[]>('/tags/first', {
      params: { size },
    });
    return response.data;
  },

  async createTag(data: TagCreateRequest): Promise<TagResponse> {
    const response = await apiClient.post<TagResponse>('/tags', data);
    return response.data;
  },

  async updateTag(id: number, data: TagUpdateRequest): Promise<TagResponse> {
    const response = await apiClient.put<TagResponse>(`/tags/${id}`, data);
    return response.data;
  },

  async deleteTag(id: number): Promise<void> {
    await apiClient.delete(`/tags/${id}`);
  },
};

